File: Algorithms/p207_Course_Schedule/p207_Course_Schedule_2.py
class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        graph = [[] for i in range(numCourses)]
        indegree = [0]*numCourses
        for e in prerequisites:
            graph[e[1]].append(e[0])
            indegree[e[0]] += 1
        q = collections.deque()
        for i in range(numCourses):
            if indegree[i]==0:
                q.append(i)
        num = 0
        while q:
            i = q.popleft()
            num += 1
            for j in graph[i]:
                indegree[j] -= 1
                if indegree[j]==0:
                    q.append(j)
        return num==numCourses

File: Algorithms/p207_Course_Schedule/p207_Course_Schedule_2.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<pair<int, int>>& prerequisites) {
        vector<vector<int>> graph(numCourses);
        vector<int> indegree(numCourses,0);
        for(auto e : prerequisites){
            graph[e.second].push_back(e.first);
            ++indegree[e.first];
        }
        queue<int> q;
        for(int i=0;i<numCourses;++i){
            if(indegree[i]==0) q.push(i);
        }
        int num = 0;
        while(!q.empty()){
            int i = q.front();
            q.pop();
            ++num;
            for(int j : graph[i]){
                if(--indegree[j]==0) q.push(j);
            }
        }
        return num==numCourses;
    }
};
